### ISSUE TITLE
Add `dtypes` parameter to `DataFrame` initialization

### ISSUE THREAD

**Issue by stinodego** [pola-rs/polars#6209]
Currently, you can enforce the dtypes of a `DataFrame` upon creation in two ways:

* Pass a list of (column, dtype) pairs to the `columns` argument
* Pass a dict of the form {column: dtype} to the `columns` argument

There are a few shortcomings of this approach:

* Specifying dtypes for a new DataFrame is unnecessarily verbose, mostly in cases where column names are inferred from the data (see examples below).
* The columns argument now has too many responsibilities, which leads to confusing behaviour (see below).
* It's not intuitive to use an argument called `columns` for specifying datatypes

I want to be able to do the following:

### Specify the dtype for only a subset of columns, and auto-infer the others:

```python
# Use a UInt8 for column 'a' - auto-infer the others
df = pl.DataFrame({'a': [1,2,3], 'b': ['x', 'y', 'z']}, dtypes={'a': pl.UInt8})
```

Technically, I can do the following to explicitly state which dtypes are auto-inferred, but it's verbose (requires to me repeat all columns):

```python
df = pl.DataFrame({'a': [1,2,3], 'b': ['x', 'y', 'z']}, columns={'a': pl.UInt8, 'b': None})
```

This is fine for a two-column dataframe, but gets old fast

### Specify dtypes without repeating column names:

```python
# Specify dtypes positionally
df = pl.DataFrame({'a': [1,2,3], 'b': ['x', 'y', 'z']}, dtypes=[pl.UInt8, pl.Utf8])
```

Currently, I need to supply a columns argument, even though I have already supplied the column names in the data dictionary:

```python
df = pl.DataFrame({'a': [1,2,3], 'b': ['x', 'y', 'z']}, columns={'a': pl.UInt8, 'b': pl.Utf8})
```

### Confusing behaviour of columns argument

This is currently valid, but quite confusing. Should we map the names from the data to a dtype, or overwrite column names and dtypes?

```python
# This overwrites names
df = pl.DataFrame([pl.Series('a', [1,2,3]), pl.Series('b', ['x', 'y', 'z'])], columns={'b': pl.UInt8, 'a': pl.Utf8})

# This maps and changes column order (added at my request by Mr. Beedie)
df = pl.DataFrame({'a': [1,2,3], 'b': ['x', 'y', 'z']}, columns={'b': pl.Utf8, 'a': pl.Int32})
```

### Implementation

We could initially implement this without it being a breaking change. The `columns` argument keeps doing what it currently does, and the added `dtypes` argument is opt-in for people who want to specify their DataFrames with strict schemas in a less verbose way.

### Long story short

We need a `dtypes` parameter because the current `columns` parameter is confusing and not concise to use when data is already named (dict or Series input).

Open for feedback on this one!

**Comment by thomasfrederikhoeck** [comment]
If looking away from breaking changes it think the best solutions is as follows:

For dtypes allow both `dict` and `list` since they don't alter names
```python
# Use a UInt8 for column 'a' - auto-infer the others
df = pl.DataFrame({'a': [1,2,3], 'b': ['x', 'y', 'z']}, dtypes={'a': pl.UInt8})

# Full specified types using list
df = pl.DataFrame({'a': [1,2,3], 'b': ['x', 'y', 'z']}, dtypes=[pl.UInt8, pl.Utf8])

# Fails since underspecified len(df.columns) != len(dtypes)
df = pl.DataFrame({'a': [1,2,3], 'b': ['x', 'y', 'z']}, dtypes=[pl.UInt8])
``` 
For columns only allow `lists` to make the sequential meaning clear. This can either be a list of strings or a list of tuples of strings and types i.e.

```python
# Use auto-infer dtypes and rename columns
df = pl.DataFrame({'a': [1,2,3], 'b': ['x', 'y', 'z']}, columns=["a_new","b_new"])

# Full specified types and names using list
df = pl.DataFrame({'a': [1,2,3], 'b': ['x', 'y', 'z']}, columns=[("a_new",pl.UInt8), ("b_new",pl.Utf8)])

# Fails since underspecified len(df.columns) != len(dtypes)
df = pl.DataFrame({'a': [1,2,3], 'b': ['x', 'y', 'z']}, columns=["a_new"])
``` 
Another alternatively is to deprecate `columns` completely and use `.rename()` together with `dtypes`. Would be nice to know how often people create series with names and rename immediately when creating a dataframe. I feel this would not be very common but just assumptions :-D

**Comment by stinodego** [comment]
That makes sense as a next step. Still, there will remain some overlap between the two parameters, which makes me a little unsure on whether this is right direction to go.

**Comment by alexander-beedie** [comment]
Hmm... I'd say the single best reason to add a standalone `dtypes` parameter is the poor ergonomics of overriding just one or two types out of a larger set of columns - that is unquestionably suboptimal! Everything else doesn't _really_ seem to be an issue; or, at least, not one that "solving" it by having overlapping options in _two_ places instead of one doesn't (potentially) make worse...🤔 

My own preference would be to keep `columns` as-is, offering `dtypes` specifically for overriding one (or more, or all) types; that way `columns` is still responsible for all things column-related (their order, names, and types _if_ present), and a `dtypes` param would _only_ be responsible for adding extra typing information. It wouldn't try to influence order, etc.

In this scenario the `dtypes` param would also _always_ take precedence over `columns`; if no type is defined in `columns` then the one in `dtypes` is used. If a type _is_ present in `columns`, it is _overridden_ by `dtypes`. No ambiguity - if you have defined something in `types`, it wins.

This gives a pretty clear delineation of responsibility, and less confusion/overlap.
* The `columns` param controls _columns_ (which _may_ include a dtype).
* The `dtypes` param controls _types_, overriding/supplementing anything in `columns`.

I'd say that to avoid more ambiguity the new `dtypes` param should actually _always_ be a dict: _"**this** col has **this** type"_. If you make it a list then you get into questions of trying to map the order of the list to keys in a dict or whatever else is going on in `columns` - at which point you would have been better off just passing your name/type info into `columns` together, so that each name/type pair is directly associated, rather than being indirectly inferred.

This would completely solve the main problem (poor ergonomics of supplying _partial_ type information), without having to put things in two places that otherwise do live pretty comfortably in one.

**Comment by ritchie46** [comment]
> I'd say that to avoid more ambiguity the new dtypes param should actually always be a dict: "this col has this type". I

If we do this I think we should name it `schema`. That is consistent with what we use in other places and I want that to feel like a natural data structure accompanying a DataFrame.

**Comment by stinodego** [comment]
> > I'd say that to avoid more ambiguity the new dtypes param should actually always be a dict: "this col has this type". I
> 
> If we do this I think we should name it `schema`. That is consistent with what we use in other places and I want that to feel like a natural data structure accompanying a DataFrame.

But then it wouldn't make sense for the use case of specifying the dtype for a single column. Maybe I'm making too big a deal of that case though - pyspark doesn't have that, for example.

I like the pyspark API though, which uses `schema` in a similar fashion to our `columns`:
https://spark.apache.org/docs/3.1.1/api/python/reference/api/pyspark.sql.SparkSession.createDataFrame.html

**Comment by ritchie46** [comment]
We already allow a partial schema. So where the schema is defined it overrules. The others are left to auto inference.

**Comment by gab23r** [comment]
A `schema` argument would be super useful especially to create an empty dataframe from the schema of another one

**Comment by alexander-beedie** [comment]
>> If we do this I think we should name it schema. 

The concept of a schema is all-encompassing though; it doesn't get you the partial aspect (which -imho- is the only part that's actively unergonomic right now). Our `columns` _is_ definitely `schema` in disguise, but I actually somewhat prefer the name we have as "schema" to me (probably from staring at too many databases over the years) strongly implies `name + type` whereas `columns` does not - it's just how you define your columns, which may _or may not_ have a type?

I could make a `Draft` PR of what this would look like in practice, and we could poke at it to see how it feels?

**Comment by alexander-beedie** [comment]
> A `schema` argument would be super useful especially to create an empty dataframe from the schema of another one
> This is super useful to create mwe of bugs for exemple 😁

You can already do this with `columns` ;)

**Comment by gab23r** [comment]
Cool! I didn't known that, I add an example in the doc  #6225

**Comment by alexander-beedie** [comment]
This turned into a two cups of coffee _and_ two cups of tea job - sometimes I forget quite how many constructor permutation we have when you look at them in aggregate...🤣 Anyway, PR will be landing shortly, for experimentation and comments.

**Comment by stinodego** [comment]
I've given this more thought, and I want to propose something a little different.

The main problem with my proposal for a `dtypes` argument is that there is overlapping responsibility between the `columns` and the `dtypes` parameters. I think this is counterintuitive and will lead to a worse user experience than before.

What about the following:

We stick to a single parameter, which is responsible for managing both column names and dtypes. It's now named `columns`, I would prefer it be renamed to `schema`. The functionality is as follows for various input types:

### List input

List input behaves exactly as it does now. It requires the user to explicitly specify schema information for each column (passing `None` to defer to auto-inferral).

  * Input is a list of either strings or string-dtype tuples
  * Overwrites names/dtypes present in the input data
  * List length must match data shape
  * If no dtype is specified for a column, the dtype of the input data is respected, or it is auto-inferred
*  _Possible addition:_ If no name is specified for a column (`None` is given), the name of the input data is respected, or it is auto-assigned:

```python
pl.DataFrame(
    [pl.Series("a", [1, 2]), pl.Series("b", [3, 4]), pl.Series("c", [5, 6])],
    columns=[("x", pl.Int32), None, (None, pl.Int8)],
)
```
```
shape: (2, 3)
┌─────┬─────┬─────┐
│ x   ┆ b   ┆ c   │
│ --- ┆ --- ┆ --- │
│ i32 ┆ i64 ┆ i8  │
╞═════╪═════╪═════╡
│ 1   ┆ 3   ┆ 5   │
│ 2   ┆ 4   ┆ 6   │
└─────┴─────┴─────┘
```

```python
pl.DataFrame(
    [[1, 2], [3,4], [5,6]],
    columns=[("x", pl.Int32), None, (None, pl.Int8)],
)
```
```
shape: (2, 3)
┌─────┬──────────┬──────────┐
│ x   ┆ column_1 ┆ column_2 │
│ --- ┆ ---      ┆ ---      │
│ i32 ┆ i64      ┆ i8       │
╞═════╪══════════╪══════════╡
│ 1   ┆ 3        ┆ 5        │
│ 2   ┆ 4        ┆ 6        │
└─────┴──────────┴──────────┘
```

### Dictionary input

Dictionary input is a convenient way to specify schema information when the data already contains some of the desired schema. Basically, it will be a shorthand for the list input, if your list input would contain a lot of `None` values. This will differ slightly from the existing functionality.

* Input is a mapping of string to dtype.
* The mapping does not need to contain a mapping for each column; a subset or empty mapping is fine.
* The names specified in the mapping MUST be present in the input data.
  * _SPECIAL CASE:_ If the input does not contain ANY names, and the number of items in the mapping equals the number of columns in the data, the column names (and dtypes) will be set to the names specified in the mapping.
* The ordering in the input data will always be respected (basically, revert #6045 - this was a bad suggestion by myself which leads to confusing behaviour).

### Implementation

The above will mean a very minor change in the existing behaviour (revert #6045 which was only added very recently). Otherwise, existing behaviour is respected, and augmented with the ability to specify only a subset of columns when passing a mapping. Also, list behaviour would be augmented with the option to specify `None` for column names.

Renaming `columns` to `schema` is a separate issue/discussion. But that wouldn't break anything - we simply deprecate `columns` in favor of `schema`.

**Comment by alexander-beedie** [comment]
Hmm... Determining the behaviour of an argument in a non-obvious way based on its type seems a bit...  pandas? 😅

If I'm reading this right you'd have one param called `schema` that when passed...
* ... as `list` it behaves one way ...
* ... and as `dict` it behaves another, non-equivalent way.
 
Having one param makes it a lot harder for the caller to say something like _"here is may standard schema"_, but also allow optional/selective overrides (which is a common pattern). Everyone who wants to do this would have to manage their own merge of the two parts every time.

However, I also agree that the param names as they are now overlap a bit too much; I do think it is functionally _much_ cleaner to have two params so as not to conceptually overload one param with different behaviours. How about we have `schema` and `overrides` - and then we lock down the behaviour of each as you suggest, getting a bit stricter on their separate responsibilities.

This way the `schema` param _always_ declares the schema. It doesn't _sometimes_ declare the schema. And if you want to override anything, you look to `overrides`.

This would also be a relatively smooth transition from the new PR; we'd add `@deprecate_alias` entries for `columns` and `dtypes`, and in the new param implement the slightly stricter separation of responsibilities.

Best of both worlds?

**Comment by stinodego** [comment]
I agree with what you're saying. I think your suggestion is really good, let me ponder it for a moment!

**Comment by stinodego** [comment]
Right, so I agree that having separate parameters is more explicit and should be preferred. We should name the 'main' parameter `schema`. I'm not sure if `overrides` is the best name for the other parameter, because it's not immediately clear what we're overriding. `schema_overrides` would probably be better, although a bit verbose... Maybe there's something even better? Either is probably fine.

Regarding the implementation: I would be against introducing a `dtypes` parameter that we then immediately deprecate and rename to `override` (or whatever that parameter will be named). Since we're introducing a new parameter, we can get it right straight away.

For `columns`, we can indeed add `@deprecated_alias` (great util by the way) for the rename to schema, and add a `FutureWarning` to the specific functionality we're changing. I think that is only the #6045 revert, right? I do not expect many people to be utilizing this functionality at the moment since it's very new, so I believe we can get rid of it quite quickly in the next breaking release.

We should keep the `@deprecated_alias` for `columns` for a good while though, since it's VERY widely used (keep it for let's say a few breaking releases).

**Comment by alexander-beedie** [comment]
@stinodego: Sounds like a plan. I think there aren't so many things that we _could_ be overriding at init time, so maybe `overrides` is actually ok? Otherwise I can't currently think of anything much shorter than `schema_overrides`, which doe feel a _bit_ unwieldy (though does have the advantage of being very explicit).

@ritchie46: Want to weigh-in on the param name? ;)

* `overrides`: shorter, readable; not much in the init param that could be overridden _except_ for dtypes?
* `schema_overrides`: very clear, obviously tied directly to `schema`, maybe a bit long?
* `<something_else>`: another suggestion that's both short and clear...?

**Comment by ritchie46** [comment]
I favour `schema_overrides`. I understand what it does immediately. And I am happy we are going for the `schema` name. :)

**Comment by alexander-beedie** [comment]
Clarity wins! I'll make the change - though possibly when I get to the airport in about two hours or so - am in the final stages of suitcase packing right now :)

**Comment by stinodego** [comment]
Enjoy the trip! Can you tag me in the PR when you feel it's ready to go?

**Comment by alexander-beedie** [comment]
@stinodego: Will do - shall I add the `deprecated_alias` to `columns` now too? Or shall we two-stage it?

**Comment by stinodego** [comment]
I'd say two-stage it, although I don't think it matters too much.

**Comment by stinodego** [comment]
This has been implemented.
